programs/vkd3d-hlsl-build/dxil_library.cpp

```diff
@@ -0,0 +1,199 @@
+
+    if (FAILED(hr))
+    {
+        std::wcerr << "Failed to create IDxcUtils instance." << std::endl;
+        return false;
+    }
+
+    IDxcCompiler3 *compiler = nullptr;
+    hr = m_dxc_create_instance(CLSID_DxcCompiler, IID_PPV_ARGS(&compiler));
+
+    if (FAILED(hr))
+    {
+        std::wcerr << "Failed to create IDxcCompiler3 instance." << std::endl;
+        utils->Release();
+        return false;
+    }
+
+    IDxcIncludeHandler *include_handler = nullptr;
+    hr = utils->CreateDefaultIncludeHandler(&include_handler);
+
+    if (FAILED(hr))
+    {
+        std::wcerr << "Failed to create include handler." << std::endl;
+        utils->Release();
+        compiler->Release();
+        return false;
+    }
+
+    /* Set up parameters that minimize the generated binary size */
+    IDxcCompilerArgs *args = nullptr;
+
+    std::array<LPCWSTR, 2> arguments = {
+        L"-Qstrip_debug",
+        L"-Qstrip_reflect",
+    };
+
+    hr = utils->BuildArguments(source_name, entry_point, target,
+            arguments.data(), arguments.size(), nullptr, 0, &args);
+
+    if (FAILED(hr))
+    {
+        std::wcerr << "Failed to build compiler arguments." << std::endl;
+        include_handler->Release();
+        utils->Release();
+        compiler->Release();
+        return false;
+    }
+
+    IDxcResult *result = nullptr;
+
+    /* Assume that shader code is some sort of single-byte encoding */
+    DxcBuffer source_buffer = { };
+    source_buffer.Ptr = code.data();
+    source_buffer.Size = code.size();
+    source_buffer.Encoding = DXC_CP_UTF8;
+
+    hr = compiler->Compile(&source_buffer,
+        args->GetArguments(), args->GetCount(),
+        include_handler, IID_PPV_ARGS(&result));
+
+    /* Compile will still return S_OK if compilation failed, we need to ask
+     * the result object for the status of the operation, and for some bizarre
+     * reason that call returns a HRESULT itself, so we have three layers of
+     * error checking here. Just make sure we end up with an error status at
+     * the end if anything fails here. */
+    if (SUCCEEDED(hr))
+    {
+        HRESULT status = S_OK;
+        hr = result->GetStatus(&status);
+
+        if (SUCCEEDED(hr))
+            hr = status;
+    }
+
+    bool status = SUCCEEDED(hr);
+
+    if (status)
+    {
+        /* Compilation successful, just retrieve the binary blob and return it. */
+        IDxcBlob *binary_blob = nullptr;
+
+        if (result && result->HasOutput(DXC_OUT_OBJECT))
+        {
+            hr = result->GetOutput(DXC_OUT_OBJECT, IID_PPV_ARGS(&binary_blob), nullptr);
+
+            if (SUCCEEDED(hr))
+            {
+                binary.resize(binary_blob->GetBufferSize());
+                std::memcpy(binary.data(), binary_blob->GetBufferPointer(), binary.size());
+            }
+        }
+
+        if (binary_blob)
+            binary_blob->Release();
+        else
+            std::wcerr << "Failed to retrieve shader binary." << std::endl;
```

Comment:
wcerr?

Reply:
We kind of need to use the `w` variants here since a lot of the strings we get are UTF-16 (including `std::filesystem::path`), and AFAIK on Windows this *needs* to be consistent throughout the application so we can't mix and match wcerr and cerr etc.

Reply:
But we're passing an ASCII string to wcerr. Is that supposed to handle the conversion for us?

Reply:
It does at least appear to work whereas the other way around (feeding wchar_t* into cerr) does not.